feat: ADR for design of a License Credential

SenorMonito opened the pull request.

noandrea commented on docs/Explanation/ADR/adr-003-issuer.md:
```suggestion
The Regulator and the Issuer Decentralized Identifier Document (DID) SHALL HAVE multiple DID controllers. For an Issuer these represent the different functions within a financial insitutions - Operations, Compliance etc. This SHALL support multiple signing such that `Pause/Unpause` requires two parties to make the change. This would mimic a case where Operations would request the pause and Compliance or a Senior Manager would approve the action.
```

noandrea commented on docs/Explanation/ADR/adr-003-issuer.md:
```suggestion
* The Regulator DID document SHALL BE defined in Genesis
```

noandrea commented on docs/Explanation/ADR/adr-005-license-credential.md:
```diff
@@ -0,0 +1,380 @@
+
+* Mint or burn e-money tokens
+* Operate a Decentralized Exchange 
+* etc. 
```
```suggestion

```

noandrea commented on docs/Explanation/ADR/adr-005-license-credential.md:
```diff
@@ -0,0 +1,380 @@
+
+## Context
+
+[ADR-003 defines an issuer module](https://github.com/allinbits/cosmos-cash/blob/main/docs/Explanation/ADR/adr-003-issuer.md). One of the actors defined in that ADR was an issuer. This issuer would hold a license that gives permissions to perform certain activities. The license is issued by a authority actor.
```
```suggestion
[ADR-003 defines an issuer module](https://github.com/allinbits/cosmos-cash/blob/main/docs/Explanation/ADR/adr-003-issuer.md). One of the actors defined in that ADR was an issuer. This issuer would hold a license that gives permissions to perform certain activities. The license is issued by a regulator actor.
```

noandrea replied:
```suggestion
[ADR-003 defines an issuer module](https://github.com/allinbits/cosmos-cash/blob/main/docs/Explanation/ADR/adr-003-issuer.md). One of the actors defined in that ADR was an issuer. This issuer would hold a license that gives permissions to perform certain activities. The license is issued by an authority actor.
```

noandrea commented on docs/Explanation/ADR/adr-005-license-credential.md:
```diff
@@ -0,0 +1,380 @@
+### Terminology
+
+* `Legal Person`: any person or entity that is able to perform legal activties, such as enter into contracts, own property, and so on
+* `Virtual Asset Service Provider`: FATF defines a VASP as the following: *“Any natural/legal person who ...  as a business conducts one or more of the following activities or operations for or on behalf of another natural or legal person: i. exchange between virtual assets and fiat currencies; ii. exchange between one or more forms of virtual assets; iii. transfer of virtual assets; iv. safekeep ...
```
source needed?

SenorMonito replied:
Yes, I can add this.

SenorMonito replied:
This section has been added to the glossary and a citation added for VASP definition.

noandrea commented on docs/Explanation/ADR/adr-005-license-credential.md:
````diff
@@ -0,0 +1,380 @@
+
+Example credential in `json-ld` format is as follows:
+
+```json
````
```suggestion
```javascript
```

noandrea commented on docs/Explanation/ADR/adr-005-license-credential.md:
```diff
@@ -0,0 +1,380 @@
+		"licenses": [{
+			"licenseType": "VQFSRO",
+			"country": "CH",
+			"authority": "Financial Services Standards Association (VQF)" # should authority be a DID?
```
```suggestion
			"authority": "Financial Services Standards Association (VQF)" // should authority be a DID?
```

noandrea commented on docs/Explanation/ADR/adr-005-license-credential.md:
```diff
@@ -0,0 +1,380 @@
+			"country": "CH",
+			"authority": "Financial Services Standards Association (VQF)" # should authority be a DID?
+		}, {
+			"licenseType": "MICAEMI", # here is a MiCA EMI license, for example
```
```suggestion
			"licenseType": "MICAEMI", // here is a MiCA EMI license, for example
```

noandrea commented on docs/Explanation/ADR/adr-005-license-credential.md:
```diff
@@ -0,0 +1,380 @@
+		"created": "2021-08-01T15:23:24Z",
+		"proofPurpose": "assertionMethod",
+		"verificationMethod": "https://example.edu/issuers/keys/1",
+		"jws": "eyJhbGciOiJSUzI1NiIsImI2NCI6ZmFsc2UsImNyaXQiOlsiYjY0Il19..." # truncated for readability
```
```suggestion
		"jws": "eyJhbGciOiJSUzI1NiIsImI2NCI6ZmFsc2UsImNyaXQiOlsiYjY0Il19..." // truncated for readability
```

SenorMonito reviewed (commented):
Thanks, @noandrea. Added suggestions

SenorMonito commented:
Additional changes to this ADR:
1. Added the architecture build and folder so I can include diagrams in the ADR. This section also had READMEs etc.
2. Split the credential ADR into two: license and registration.
3. Other chore changes.

barriebyron commented on docs/Explanation/ADR/adr-006-license-credential.md:
```diff
@@ -0,0 +1,217 @@
+This ADR offers the following benefits:
+
+- Regulatory compliance through open standards
+- It implements a Role Based Access to services on Cosmos Cash.
+- It uses Self-Sovereign Identity (SSI) standards so this means:
```
```suggestion
- Implements a Role Based Access to services on Cosmos Cash.
- Uses Self-Sovereign Identity (SSI) standards so this means:
```

barriebyron commented on docs/Explanation/ADR/adr-006-license-credential.md:
```diff
@@ -0,0 +1,217 @@
+- Regulatory compliance through open standards
+- It implements a Role Based Access to services on Cosmos Cash.
+- It uses Self-Sovereign Identity (SSI) standards so this means:
+  - Interoperability - for example the credential can be stored in a variety of different identity wallets
```
```suggestion
  - Interoperability - for example, the credential can be stored in a variety of different identity wallets
```

barriebyron commented on docs/Explanation/ADR/adr-006-license-credential.md:
```diff
@@ -0,0 +1,217 @@
+- It uses Self-Sovereign Identity (SSI) standards so this means:
+  - Interoperability - for example the credential can be stored in a variety of different identity wallets
+  - Flexibility - for example, this credential can be re-used in other scenarios where token issuance is required
+- Adoption of SSI for provision of financial services offers benefits in terms of licensing and management on public platforms.
```
we haven't talked yet about SSI, what is this? link to trusted resource or our own section?

barriebyron commented on docs/Explanation/ADR/adr-006-license-credential.md:
```diff
@@ -0,0 +1,217 @@
+There are a number of macro-level concerns, but these do not affect the adoption of this model for Role Based Access Control (RBAC). These macro concerns include:
+
+- Self-Sovereign Identity technologies are still relatively new. 
+- eiDAS legislation is still in development.
```
what is eiDAS legislation? can we link to a trusted resource?

barriebyron commented on docs/Reference/GLOSSARY.md:
```diff
@@ -0,0 +1,41 @@
+# GLOSSARY
+
```
would love to help develop this glossary! 
For style only (not substantive use or description) we are using this organic [Word List Cosmo](https://docs.google.com/spreadsheets/d/15raX8lfzzeeifGrDiXSFTHaEDbsMVlFJzKgnNelihjs/edit#gid=0) in partnership with B9lab (the consulting firm doing a learning portal for Cosmos SDK). Part of https://github.com/cosmos/advocates/issues/23 in the Cosmos SDK advocates repo

barriebyron commented on docs/Reference/architecture/CONTRIBUTING.md:
```diff
@@ -0,0 +1,84 @@
+
+## Installation
+
+Cosmos Cash diagrams are based on PlantUML. 
```
based on? or developed using?

barriebyron commented on docs/Reference/architecture/CONTRIBUTING.md:
```diff
@@ -0,0 +1,84 @@
+
+Linux-based systems, through your distributions package manager
+
+## Contributing diagrams
```
let's use our standard Title case style capitalization for all headings, I did not not all of them below, it's handy to use https://capitalizemytitle.com/

barriebyron commented on docs/Reference/architecture/CONTRIBUTING.md:
```diff
@@ -0,0 +1,84 @@
+
+## Contributing diagrams
+
+The source code for all diagrams SHOULD be stored in `./src` folder AND have the `.puml` suffix. Please refer to the [PlanUML documentation](https://plantuml.com) for further details.
```
```suggestion
The source code for all diagrams SHOULD be stored in `./src` folder AND have the `.puml` suffix. See the [PlanUML documentation](https://plantuml.com) for further details.
```

barriebyron replied:
we don't want to use Please, we already are friendly and helpful, per https://developers.google.com/style/tone#politeness-and-use-of-please

barriebyron commented on docs/Reference/architecture/CONTRIBUTING.md:
```diff
@@ -0,0 +1,84 @@
+
+## Themes
+
+Where possible all diagrams will be rendered to have a similar look and feel. To achieve this add the following to your diagram:
```
```suggestion
Where possible all diagrams are rendered to have a similar look and feel. To achieve this look and feel, add the following theme to your diagram:
```

barriebyron replied:
now that we're out of the ADR (requires future tense, per Cosmos SDK convention) let's use standards for present tense, per https://developers.google.com/style/tense

barriebyron commented on docs/Reference/architecture/CONTRIBUTING.md:
```diff
@@ -0,0 +1,84 @@
+
+### PlantUML CLI
+
+There is [makefile](.makefile) that will build all `src/*.puml` files. To create SVG images just run `make (svg|png)` as needed. `make clean` will get a clean folder ready for new build.
+
+This is the preferred and cleanest way to create all the architecture documentation.
```
```suggestion
The [makefile](.makefile) builds all `src/*.puml` files. To create SVG images, run `make (svg|png)` as needed. Run `make clean` to get a clean folder ready for a new build.

Using PlantUML and makefile is the preferred and cleanest way to create all the architecture documentation.
```

barriebyron commented on docs/Reference/architecture/CONTRIBUTING.md:
```diff
@@ -0,0 +1,84 @@
+
+This method uses PlantUML's [proxy service](https://plantuml.com/server) to put generate images and embed into a Markdown page. 
+
+To use the proxy service integration simply use:
```
```suggestion
To use the proxy service integration, use:
```

barriebyron replied:
it's tempting when we think something is easy, but let's omit "simply" in all cases, per https://developers.google.com/style/tone?hl=en#some-things-to-avoid-where-possible

barriebyron commented on docs/Reference/architecture/CONTRIBUTING.md:
```diff
@@ -0,0 +1,84 @@
+
+### Visual Studio Code
+
+If you use [Visual Studio Code](https://code.visualstudio.com/) there is [PlantUML plugin](https://marketplace.visualstudio.com/items?itemName=jebbs.plantuml) available. 
+
+From here, one can preview diagrams through the command palette (CTRL+SHIFT+P on Windows/Linux or COMMAND+SHIFT+P on MacOS)
```
```suggestion
If you use [Visual Studio Code](https://code.visualstudio.com/), a [PlantUML plugin](https://marketplace.visualstudio.com/items?itemName=jebbs.plantuml) is available. 

From here, you can preview diagrams by using the command palette (CTRL+SHIFT+P on Windows/Linux or COMMAND+SHIFT+P on macOS).
```

barriebyron commented on docs/Reference/architecture/README.md:
```diff
@@ -0,0 +1,12 @@
+
+## Overview
+
+Architecture documentation is based on [PlantUML](https://plantuml.com/). This contains lots of documentation and examples. Diagrams are created through text based documents (`.puml` files) which are then rendered into an appropriate format, in a similar manner to pandoc for markdown and pdfs/html.
```
```suggestion
Architecture documentation is based on [PlantUML](https://plantuml.com/) that comes with lots of documentation and examples. Diagrams are created through text-based documents (`.puml` files) that are then rendered into an appropriate format like Pandoc Markdown is used to generate PDFs and HTML output.
```

barriebyron commented on docs/Reference/architecture/README.md:
```diff
@@ -0,0 +1,12 @@
+1. It supports a number of different types of diagram: class, sequence, activity, ERM ...
+2. Diagrams can be stored with the codebase and documentation rather in a separate system such as Whimsical and then imported as needed into documentation.
+
+To contribute please see [CONTRIBUTING](CONTRIBUTING.md)
```
```suggestion
To contribute, see [CONTRIBUTING](CONTRIBUTING.md).
```

barriebyron reviewed (commented):
I always have a great time reviewing this type of content. I suggested many (oof) granular changes to comply with styles, ensure consistency, and apply best practices for language and clarity. 
lmk if this level of review is helpful. I could check out the branch and make direct changes, but IMO providing line-based suggestions with links and resources and questions works well for this type of content.

Have fun with it all! lmk how I can do better. And make today an awesome day.

PaddyMc commented on docs/Explanation/ADR/adr-006-license-credential.md:
```diff
@@ -0,0 +1,217 @@
+
+Financial services regulation is based on registrations issued by a publicly appointed authority for a given jurisdiction. Registrations together with licenses allow legal entities to perform activities such holding client money, offer financial advice etc. Typically a regulator will maintain a central public registry, for example the [FCA register](https://register.fca.org.uk/s/). 
+
+This ADR describes the structure of an issuer license based on the [W3C specification](https://www.w3.org/TR/vc-data-model/) for verifiable credential. This license makes reference to the classes of crypto assets described in MiCA, but the license could easily be adopted as proof of authority to issue various types of crypto or virtual assets. In all cases this credential will be signed by an `au ...
```
Should we specify these are on-chain credentials?

PaddyMc reviewed (commented):
Great first draft, I've been through the license credentials ADR and created corresponding proto files.

Please have a look at this PR for reference => https://github.com/allinbits/cosmos-cash/pull/198/files

Will review the registration credential now :bow:

PaddyMc commented on docs/Explanation/ADR/adr-005-registration-credential.md:
```diff
@@ -0,0 +1,332 @@
+				"country": "GB"
+			},
+			"ids": [{
+				"idId": "CH02035398520",
```
Can we name this something more concise? Maybe just `id` is fine here?

PaddyMc commented on docs/Explanation/ADR/adr-005-registration-credential.md:
```diff
@@ -0,0 +1,332 @@
+			},
+			"ids": [{
+				"idId": "CH02035398520",
+				"idType": "RAID"
```
Maybe this can just be called `type` as well?

PaddyMc commented on docs/Explanation/ADR/adr-005-registration-credential.md:
```diff
@@ -0,0 +1,332 @@
+| Level 3   | Name        | Type                    | Mult. | Notes     |
+| :-------- | :---------- | :---------------------- | :---- | :-------- |
+| Name      | `name`      | String                  | 1..1  |           |
+| Name type | `nameType`  | [Name Type](#name-type) | 1..1  |           |
```
Maybe just `type` is fine here as well? wdyt?

PaddyMc reviewed (commented):
Looks good to me so far, most of my comments on the registration cred are to do with naming, e.g `idId` => `id`

I sketched the proto files for this ADR as well, they can be seen here => https://github.com/allinbits/cosmos-cash/pull/199

Please review and lmk! :bow:

SenorMonito commented on docs/Explanation/ADR/adr-003-issuer.md:
```diff
@@ -0,0 +1,154 @@
+* Rescue funds from frozen accounts
+* Pause/Unpause token circulation
+
+This functionality will use Role Based Access based on Verifiable Credentials with blocklists maintained through the use of revocation lists.
```
response: yes, we should use RBAC here.
